### PR TITLE
Add rate aggregation

### LINKED ISSUE #60674: Add new rate aggregation
We have received several requests to add "rate" functionality or aggregation: how many docs/s were there per bucket in a date_histogram? We have decided to implement this as a special metric aggregation with a scope limited to only `date_histogram` aggregation at the moment, in other words, the `rate` aggregation will have to be a descendant of a `date_histgram` and in the case of nested histograms the closest ancestor will be used to determine the rate. 

```
POST /sales/_search
{
  "aggs": {
    "sales": {
      "date_histogram": {
        "field": "date",
        "calendar_interval": "day"
      },
      "aggs": {
        "the_rate": {
          "rate": {
            "unit": "second" 
          } 
        }
      }
    }
  }
}

{
  ...
  "aggregations": {
    "sales": {
      "buckets": [
        {
          "key_as_string": "2020-07-29",
          "doc_count": 300000,
          "the_rate": {
            "rate": 3.47222222222,
            "rate_as_string": "3.47222222222/s"
          }
        },
       ...
      ]
    }
  }
}
```

By default the number of documents in the bucket will be used to calculate the rate, but it will be also possible to specify a numeric field to use sum of all values of the field to calculate the rate:

```
      "aggs": {
        "the_rate": {
          "rate": {
            "field": "num_of_requests" 
            "unit": "second" 
          } 
        }
      }
```

We could also add support for `"accumulative" : true`  flag to address #60619 in a future iteration.

### PR DESCRIPTION
**Pull request by imotov**
Adds a new rate aggregation that can calculate a document rate for buckets
of a date_histogram.

Closes #60674

### CODE REVIEW

**Review by not-napoleon — APPROVED** [review]
Overall, LGTM.  The nits are ignorable, but the comments about null values source/create unmapped, and about supporting DATE values should be addressed before merging.  Thanks!

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/RateAggregationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregationBuilder;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.aggregations.support.ValuesSourceRegistry;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+public class RateAggregationBuilder extends ValuesSourceAggregationBuilder.LeafOnly<ValuesSource.Numeric, RateAggregationBuilder> {
```
[the remark this answers is not in the document]
Comment (imotov) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/RateAggregationBuilder.java:31]: :+1:

**Comment by imotov** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Review by csoulios — APPROVED** [review]
LGTM! 

As long as the bucket `doc_count` is retrieved using the `BucketsAggregator#getDocCounts()` it will seamlessly work with aggregate doc counts!

**Comment by imotov** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Pull request merged**